planner/cascades/implementation_rules.go

```diff
@@ -436,3 +442,49 @@ func (r *ImplUnionAll) OnImplement(expr *memo.GroupExpr, reqProp *property.Physi
 	physicalUnion.SetSchema(expr.Group.Prop.Schema)
 	return impl.NewUnionAllImpl(physicalUnion), nil
 }
+
+// ImplApply implements LogicalApply to PhysicalApply
+type ImplApply struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplApply) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	return prop.AllColsFromSchema(expr.Children[0].Prop.Schema)
+}
+
+// OnImplement implements ImplementationRule OnImplement interface
+func (r *ImplApply) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+	la := expr.ExprNode.(*plannercore.LogicalApply)
+	join := la.GetHashJoin(reqProp)
+	physicalApply := plannercore.PhysicalApply{
+		PhysicalHashJoin: *join,
+		OuterSchema:      la.CorCols,
+	}.Init(
+		la.SCtx(),
+		expr.Group.Prop.Stats.ScaleByExpectCnt(reqProp.ExpectedCnt),
+		la.SelectBlockOffset(),
+		&property.PhysicalProperty{ExpectedCnt: math.MaxFloat64, Items: reqProp.Items},
+		&property.PhysicalProperty{ExpectedCnt: math.MaxFloat64})
+	physicalApply.SetSchema(expr.Group.Prop.Schema)
+	return impl.NewApplyImpl(physicalApply), nil
+}
+
+// ImplMaxOneRow implements LogicalMaxOneRow to PhysicalMaxOneRow.
+type ImplMaxOneRow struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplMaxOneRow) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	return prop.IsEmpty()
```

Comment:
In fact, max one row could satisfy any property?

Reply:
It keeps the same as the old planner. In my opinion, it should be an assertion.